### ISSUE TITLE
No automatic sync of reading progress

### ISSUE THREAD

**Issue by pbienst** [koreader/koreader#5109]
I have koreader 2019.06 both on a Klara HD and on my Android device. Opening the app on my phone automatically pulls the reading progress from my Kobo. However, when I switch to my Kobo, this is not happening automatically, and I always need to manually go into the menu and pull the reading progress...

**Comment by KenMaltby** [comment]
Perhaps KOReader on your Kobo doesn't start with the wifi on?

**Comment by Frenzie** [comment]
There shouldn't be any functional difference here since it's all high-level stuff. Like @KenMaltby said it might be different specific conditions or other subtly different (user) behavior. A quick peek at the code suggests there's auto-sync on connecting to a network, on loading a document and on resume.

Could you perhaps describe the expected and observed behavior in a little more detail?

**Comment by pbienst** [comment]
Not sure there's a lot more to say beyond what I already wrote: if I start koreader on Kobo, the reading progress is never automatically updated, even if I wait a few minutes. Only when I go into the menu option to explicitly ask a pull of the reading progress, it is correctly updated. 

No error messages related to wifi...

**Comment by Frenzie** [comment]
So you're saying wifi is enabled?

**Comment by pbienst** [comment]
Yes, it's enabled. (Unless koreader lies to me and the wifi only gets properly activated once I explicitly ask for a sync.)

**Comment by NiLuJe** [comment]
WiFi on Kobo is *extremely* finicky. If anything looks wonky at KOReader startup, we *will* kill it, to ensure you'll be able to properly control it later.

Which means the *only* way to actually boot with WiFi enabled, is.... to boot with WiFi *sanely* enabled & working.

Which leads to my next question: how are you starting KOReader, and is WiFi usually enabled at that time (and, if so, how)?

**Comment by pbienst** [comment]
OK, I did a lot more testing.

I launch koreader on my Kobo by clicking the 'koreader' book from the default os. I verified that wifi is enabled and connected to the right network, swipe a few pages in my book, turn off the Kobo. Then I turn on koreader on my phone, and it almost immediately syncs to the new page. I swipe a couple of pages on my phone, turn it off, and then turn on my Kobo again. No sync message appears. Clicking the 'gear' settings icon shows that wifi is still enabled, network info shows the expected info. 
When I click 'plugin icon - progress sync - pull progress' it says 'latest progress is coming from this device'.
Going back to my phone, I do a 'push progress from this device'. Turning the kobo off and and on still does not pick up the progress automatically. Manually pulling the progress still results in 'latest progress is coming from this device'.

However, using a different book, rather than 'latest progress is coming from this device', it does sync, but only after manually pulling the progress. 

It also turns out that for some books, it syncs automatically from phone to Kobo, but not the other way around, whereas different books or the same book at a different moment in time show the opposite behaviour.

In short, I don't think wifi is to blame, but rather there is something fishy in the sync algorithm...

**Comment by NiLuJe** [comment]
That would indeed seem to rule out WiFi shenanigans ;).

As I have absolutely zero experience with the sync thingy, I'll leave the floor to people who do ;).

**Comment by NiLuJe** [comment]
Oh, wait.

When you say "turn off the Kobo", I assume you mean "put it to sleep"?

Which means you have the "restore WiFi on resume" option enabled?

In which case, there might be a timing issue, since this is done asynchronously, and nothing really waits for it to be done before trying to do stuff over WiFi.

**Comment by Frenzie** [comment]
Looks like the connected event is sent only through the menu, so for anything else you'd need open document stuff.

https://github.com/koreader/koreader/blob/a13f9835beeca05eef57756a522a375c34f07e98/frontend/ui/network/manager.lua#L154

The resume event may come more than a second before the network actually reconnects.

https://github.com/koreader/koreader/blob/850be52177a08114013fa6af11a36f1e4081918a/plugins/kosync.koplugin/main.lua#L615-L617

So just scheduleIn a few seconds perhaps? Not foolproof, but 3-5 seconds might catch the worst of it.

**Comment by pbienst** [comment]
Yes, restore wifi is enabled.

**Comment by NiLuJe** [comment]
Ah, `restoreWifiAsync` never sends a `NetworkConnected` event, that would probably explain it ^^.

**Comment by Frenzie** [comment]
So we could just stick a scheduleIn 5 seconds or something behind it with an isConnected and then send an event?

https://github.com/koreader/koreader/blob/38e7975728ccc2a95f97632056fb710209453883/frontend/device/kobo/device.lua#L402-L404

**Comment by NiLuJe** [comment]
That was my working theory, too, yeah.

Otherwise, not quite sure how feasible it would be to do something a little b it uglier, but a little bit more fool-proof:

Schedule something in n sec, for a maximum of i iterations, that checks if the connection was successful, and send the event, or reschedule itself. Then stop after the iteration limit and/or on successful connection...

**Comment by Frenzie** [comment]
True, I suppose you could retry about twice.

**Comment by NiLuJe** [comment]
Not sure if something like this is actually valid Lua :D.

```diff
diff --git a/frontend/device/kobo/device.lua b/frontend/device/kobo/device.lua
index aa78c15e..3a100109 100644
--- a/frontend/device/kobo/device.lua
+++ b/frontend/device/kobo/device.lua
@@ -401,6 +401,24 @@ function Kobo:initNetworkManager(NetworkMgr)
 
     function NetworkMgr:restoreWifiAsync()
         os.execute("./restore-wifi-async.sh")
+
+        -- Make sure we eventually send a NetworkConnected event, as a few things rely on it (KOSync, c.f. #5109).
+        local UIManager = require("ui/uimanager")
+        local function connectivityCheck(iter)
+            -- Give up after a while...
+            if iter > 4 then
+                return
+            end
+
+            if NetworkMgr:isWifiOn() and NetworkMgr:isConnected() then
+                local Event = require("ui/event")
+                UIManager:broadcastEvent(Event:new("NetworkConnected"))
+            else
+                UIManager:scheduleIn(5, connectivityCheck(iter + 1), end)
+            end
+        end
+
+        UIManager:scheduleIn(5, connectivityCheck(1), end)
     end
 
     -- NOTE: Cheap-ass way of checking if WiFi seems to be enabled...
```

**Comment by Frenzie** [comment]
Why wouldn't it be? :-)

Btw, iirc roughly the same code lives in Cervantes and probably some others (Sony?) so if you added that for real it should be a little more generalized.

**Comment by NiLuJe** [comment]
Dunno, recursion can be weird, I guess? ^^

Yep, I was thinking of moving the actually device-specific bit to another function, and make the "public" restoreWifiAsync() a wrapper that includes this bit, so it doesn't have to be duplicated for every platform that need it.

**Comment by NiLuJe** [comment]
Well, recursion is indeed weird ;).

That fails the UIManager:scheduleIn assertion because action is nil.
Getting a reference to the current function can apparently only be done through the debug module, and it's horrendously slow (apparently). So, yay?.

**Comment by pbienst** [comment]
Are these patches supposed to be include in 2019.8? If so, I'm afraid nothing has changed...

**Comment by Frenzie** [comment]
Yes, anything that's merged in master is automatically included in the following nightly. (And eventually, the stable.)

Could you share the crash.log file?

**Comment by pbienst** [comment]
There is no crash, it just shows the behaviour I described above. As I already mentioned, I don't think wifi is the culprit, more likely it's something in the sync algorithm.

**Comment by Frenzie** [comment]
It's about the log, doesn't matter if there's a crash or not. That's just what it's called. ;)

**Comment by pbienst** [comment]
OK, I take it back: looking at the logs it indeed seems that the wifi is not yet up:

`
08/17/19-22:43:21 INFO  Kobo WiFi: disabling WiFi
08/17/19-22:43:36 INFO  Kobo suspend: going to sleep . . .
08/17/19-22:43:36 INFO  Kobo suspend: asked the kernel to put subsystems to sleep, ret: true
08/17/19-22:43:38 INFO  Kobo suspend: waited for 2s because of reasons...
08/17/19-22:43:38 INFO  Kobo suspend: synced FS
08/17/19-22:43:38 INFO  Kobo suspend: asking for a suspend to RAM . . .
08/17/19-22:43:38 INFO  Kobo suspend: ZzZ ZzZ ZzZ? Write syscall returned:  true
08/18/19-14:11:49 INFO  Kobo suspend: woke up!
08/18/19-14:11:49 INFO  Kobo resume: clean up after wakeup
08/18/19-14:11:49 INFO  Kobo resume: unflagged kernel subsystems for resume, ret: true
[Sun Aug 18 14:11:49 CEST 2019] restore-wifi-async.sh: Restarting WiFi
GET https://sync.koreader.rocks:443/syncs/progress/f247fc8c039c4f2c07929208e05aa4a7
host or service not provided, or not known
udhcpc (v1.22.0.kobo) started
Sending discover...
Sending select for 192.168.2.192...
Lease of 192.168.2.192 obtained, lease time 3600
adding dns 195.130.130.2
adding dns 195.130.131.2
[Sun Aug 18 14:11:54 CEST 2019] restore-wifi-async.sh: Restarted WiFi
PING 192.168.2.1 (192.168.2.1): 56 data bytes
64 bytes from 192.168.2.1: seq=0 ttl=64 time=3.392 ms
`

**Comment by NiLuJe** [comment]
Could we get an actual, complete log, instead of snippets?

**Comment by pbienst** [comment]
The full log is below, but note the snippet above starts from yesterday, when I turned off the device, and continues to when I woke it up today and did a manual progress pull.

https://www.dropbox.com/s/vme36lo91czow04/crash.log?dl=0

**Comment by NiLuJe** [comment]
I'm exceedingly unfamiliar with the plugin, but it looks like the onResume sync is failing (which is expected), but I'm not seeing the second one from onNetworkConnected....

I don't know if the plugin has more verbose logging in debug mode, but it's worth a try: you cna enable verbose debug logging in the "Developer settings" entry of ther FM's tool tab (restart KOreader after that, for good measure).

**Comment by pbienst** [comment]
New kobo logs with extra debug info:

https://www.dropbox.com/s/vme36lo91czow04/crash.log?dl=0

They are very strange though. The scribes book, which I've managed to sync successfully, displays some errors about service not know.

The mond book, which erroneously said "Latest sync progress already coming from this device", doesn't seem to have generated any log entry.

I've triple-checked that the filenames are identical on both devices.
